### PR TITLE
Feature/#131 Created NotificationSenderService services

### LINKED ISSUE #131: Create SMS and Email sender service and brokers
### Description
Create SMS and Email sender with necessary infrastructure

### Requirements
- project should be runnable
- sending notifications must be successful

### Deliverables
- broker external configuration model and configuration values
- `EmailMessage` and `SmsMessage` models
- `FuncResult` and exception extensions
- `EmailSenderBroker` and `SmsSenderBroker` brokers
- `EmailSenderService` and `SmsSenderService` brokers
- mapper between template and message
- validation for messages
- register all necessary infrastructure to DI

### PR DESCRIPTION
**Pull request by Ozodbek-Anvarjonov**
Closes #131

### CODE REVIEW

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs [hunk 1]
```diff
@@ -19,6 +24,35 @@ static HostConfiguration()
         Assemblies.Add(Assembly.GetExecutingAssembly());
     }
 
+
+    /// <summary>
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs:27]: Remove unnecessary row

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationType.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// indicates that the notification type is email
+    /// </summary>
+    Email = 0,
+
+    /// <summary>
+    /// indicates that the notification type is sms
+    /// </summary>
+    Sms = 1,
+}
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationType.cs:17]: Enumni propertylariga docs yozish kerakmikan?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class EmailSenderService(IEnumerable<IEmailSenderBroker> emailSenderBrokers, IValidator<EmailMessage> validator)
+    : IEmailSenderService
+{
+    private IEnumerable<IEmailSenderBroker> _mailSenderBroker => emailSenderBrokers;
+    private IValidator<EmailMessage> _emailMessageValidator => validator;
+
+
+    public async ValueTask<bool> SendAsync(EmailMessage emailMessage, CancellationToken cancellationToken = default)
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs:16]: remove unnecessary row

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class SmsSenderService(IEnumerable<ISmsSenderBroker> smsSenderBroker, IValidator<SmsMessage> smsSenderValidator)
+    : ISmsSenderService
+{
+    private IEnumerable<ISmsSenderBroker> _smsSenderBroker => smsSenderBroker;
+    private IValidator<SmsMessage> _smsSenderValidator => smsSenderValidator;
+
+
+    public async ValueTask<bool> SendAsync(SmsMessage smsMessage, CancellationToken cancellationToken = default)
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs:16]: remove unncessary row

**Review by Xolmuratov02 — APPROVED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationType.cs (hunk 2 above)
Comment (Ozodbek-Anvarjonov) [reply src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationType.cs:17]: photo_2023-11-25_21-59-21

**Review by botirjonabdumannonovv — DISMISSED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+/// <summary>
+/// Represents a notification event
+/// </summary>
+public enum NotificationEvent
+{
+    /// <summary>
+    /// status in rendering the notification
+    /// </summary>
+    OnRedering,
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs]: Typo in this one

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿namespace TruckWorld.Domain.Enums;
+
+/// <summary>
+/// Represents a notification event
+/// </summary>
+public enum NotificationEvent
+{
+    /// <summary>
+    /// status in rendering the notification
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Enums/NotificationEvent.cs]: Add "Indicates" prefix in comments for enum values

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
+﻿using TruckWorld.Domain.Common.Exceptions;
+
+namespace TruckWorld.Domain.Extensions;
+
+/// <summary>
+/// Represents any entity that return data or exetption
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs]: Change to "Provides exception extensions"

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+/// </summary>
+public static class ExceptionExtensions
+{
+    public static async ValueTask<FuncResult<T>> GetValueAsync<T>(this Func<Task<T>> func) where T : struct
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs:17]: Add comments as "Try catch wrapper for async functions"

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿using AutoMapper;
+
+namespace TruckWorld.Infrastructure.Common.Mapper;
+
+public class NotificationMessageMapper : Profile
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs:5]: Why it's empty ?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    public async ValueTask<bool> SendAsync(EmailMessage emailMessage, CancellationToken cancellationToken = default)
+    {
+        emailMessage.SenderEmailAddress ??= _settings.CredentialAddress;
+
+        var mail = new MailMessage(emailMessage.SenderEmailAddress, emailMessage.RecieverEmailAddress);
+
+        mail.Subject = emailMessage.Subject;
+        mail.Body = emailMessage.Body;
+        mail.IsBodyHtml = true;
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Brokers/SmtpEmailSenderBroker.cs:25]: We need to get this from notification template

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    : IEmailSenderService
+{
+    private IEnumerable<IEmailSenderBroker> _mailSenderBroker => emailSenderBrokers;
+    private IValidator<EmailMessage> _emailMessageValidator => validator;
+
+    public async ValueTask<bool> SendAsync(EmailMessage emailMessage, CancellationToken cancellationToken = default)
+    {
+        var validationResult = _emailMessageValidator.Validate(emailMessage,
+            options => options
+            .IncludeRuleSets(NotificationEvent.OnRedering.ToString())
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs]: Why we need `OnRendering` condition here ?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+using TruckWorld.Domain.Enums;
+using TruckWorld.Domain.Extensions;
+
+namespace TruckWorld.Infrastructure.Common.Notifications.Services;
+
+public class EmailSenderService(IEnumerable<IEmailSenderBroker> emailSenderBrokers, IValidator<EmailMessage> validator)
+    : IEmailSenderService
+{
+    private IEnumerable<IEmailSenderBroker> _mailSenderBroker => emailSenderBrokers;
+    private IValidator<EmailMessage> _emailMessageValidator => validator;
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/EmailSenderService.cs:14]: Don't need private fields, just use `emailSenderBrokers` that we got from primary constructor

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    : ISmsSenderService
+{
+    private IEnumerable<ISmsSenderBroker> _smsSenderBroker => smsSenderBroker;
+    private IValidator<SmsMessage> _smsSenderValidator => smsSenderValidator;
+
+    public async ValueTask<bool> SendAsync(SmsMessage smsMessage, CancellationToken cancellationToken = default)
+    {
+        var validationResult = _smsSenderValidator.Validate(smsMessage,
+            options => options
+            .IncludeRuleSets(NotificationEvent.OnRedering.ToString())
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Notifications/Services/SmsSenderService.cs]: Why we need OnRendering condition here ?

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Notificaitons/Brokers/ISmsSenderBroker.cs [hunk 14]
```diff
@@ -0,0 +1,17 @@
+﻿using TruckWorld.Application.Common.Notificaitons.Models;
+
+namespace TruckWorld.Application.Common.Notificaitons.Brokers;
+
+/// <summary>
+/// Represent sending sms messages
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Notificaitons/Brokers/ISmsSenderBroker.cs]: Change to "Provides methods to send SMS messages"

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Notificaitons/Brokers/IEmailSenderBroker.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
+﻿using TruckWorld.Application.Common.Notificaitons.Models;
+
+namespace TruckWorld.Application.Common.Notificaitons.Brokers;
+
+/// <summary>
+/// Represent sending email messages
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Application/Common/Notificaitons/Brokers/IEmailSenderBroker.cs]: Change to "Provides methods to send Email messages"

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs [hunk 16]
```diff
@@ -19,6 +24,34 @@ static HostConfiguration()
...
+        builder.Services
+            .AddScoped<ISmsSenderBroker, TwilioSmsSenderBroker>()
+            .AddScoped<IEmailSenderBroker, SmtpEmailSenderBroker>();
+
+        // register helper foundation services
+        builder.Services
+            .AddScoped<ISmsSenderService, SmsSenderService>()
+            .AddScoped<IEmailSenderService, EmailSenderService>();
+
+        builder.Services.AddValidatorsFromAssemblies(Assemblies);
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs]: Move to `AddValidators` method

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs [hunk 17]
```diff
@@ -19,6 +24,34 @@ static HostConfiguration()
...
+    /// <summary>
+    /// Register NotificationInfrastructure
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <returns></returns>
+    private static WebApplicationBuilder AddNotificationInfrastructure(this WebApplicationBuilder builder)
+    {
+        // register configurations 
+        builder.Services
+            .Configure<ValidationSettings>(builder.Configuration.GetSection(nameof(ValidationSettings)))
```
Comment (sulton-max) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Api/Configurations/HostConfiguration.Extensions.cs]: Move to `AddValidators` method

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs (hunk 9 above)
Comment (Ozodbek-Anvarjonov) [reply src/Training.TruckWorld.Clone.Backend/TruckWorld.Infrastructure/Common/Mapper/NotificationMessageMapper.cs:5]: boshqa model orqali map qilinadi, u model hozircha kodda mavjud emas

**Review by sulton-max — DISMISSED** [review]

**Review by sulton-max — APPROVED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+public static class ExceptionExtensions
+{
+    /// <summary>
+    /// Try catch wrapper for async functions
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="func"></param>
+    /// <returns></returns>
+
+    public static async ValueTask<FuncResult<T>> GetValueAsync<T>(this Func<Task<T>> func) where T : struct
```
Comment (botirjonabdumannonovv) [inline src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs:17]: bitta qatorni o'chiriib tashang

**Review by botirjonabdumannonovv — APPROVED** [review]

File: src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs (hunk 18 above)
Comment (Ozodbek-Anvarjonov) [reply src/Training.TruckWorld.Clone.Backend/TruckWorld.Domain/Extensions/ExceptionExtensions.cs:17]: ortiqcha qator yo'q manimcha

**Pull request merged**